File: session_4/driver/driver.go
package main

import "fmt"

type DBDriver struct {
}

type Row struct{}

func (d *DBDriver) Execute(sql string, vars []any) []Row {
	msg := d.Prepare(sql, vars)
	_ = msg
	// d.conn.send(msg)
	var out []byte
	// out := db.conn.recv()
	return d.Parse(out)
}

func (d *DBDriver) Prepare(sql string, vars []any) string {
	fmt.Println("DBDriver")
	return ""
}

func (d *DBDriver) Parse(data []byte) []Row {
	return nil
}

type PGDriver struct {
	DBDriver
}

func (d *PGDriver) Prepare(sql string, vars []any) string {
	fmt.Println("PGDriver")
	return ""
}

func (d *PGDriver) Parse(data []byte) []Row {
	return nil
}

func main() {
	drv := PGDriver{}
	drv.Execute("...", nil) // Will print DBDriver, not PGDriver
}

File: driver/driver.py
from unittest.mock import MagicMock


class DBDriver:
    def execute(self, sql, vars):
        msg = self.prepare(sql, vars)
        self.conn.send(msg)
        data = self.conn.recv()
        return self.parse(data)

    def prepare(self, sql, vars):
        raise NotImplementedError()

    def parse(self, data):
        raise NotImplementedError()


class PGDriver(DBDriver):
    def prepare(self, sql, vars):
        print('type:', type(self))
        return b''  # FIXME

    def parse(self, data):
        return []  # FIXME


drv = PGDriver()
drv.conn = MagicMock()
drv.execute('SELECT name FROM users WHERE id=?', (900,))
